### PR TITLE
SAMZA-2156: Couchbase support for Samza Table API

### PR DESCRIPTION
**Pull request by Zhangyx39**
Implemented CouchbaseTableReadFunction
Implemented CouchbaseTableWriteFunction
Tested with unit tests and integration test

### CODE REVIEW

**Comment by Zhangyx39** [comment]
@weisong44 @dengpanyin I created a new merge request against the real master instead of my forked master.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+ * @param <V> Type of values to read from / write to couchbase
+ */
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
```
Comment (rmatharu) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: I think the default serde, shouldn't be null, since a non-null valueSerde is required for working of this class.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    COUCHBASE_BUCKET_REGISTRY.closeBucket(bucketName, clusterNodes);
+  }
+
+  /**
+   * Check whether the exception is caused by one of the temporary failure exceptions, which are
+   * likely to be retriable.
+   * @param exception exception thrown by the table provider
+   * @return a boolean
+   */
+  public boolean isRetriable(Throwable exception) {
```
Comment (rmatharu) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: Seems like a helper function, which could just be a static. 
Also, maybe moved down in the class.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    documentType = JsonObject.class.isAssignableFrom(valueClass) ? JsonDocument.class : BinaryDocument.class;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info(String.format("Read function for bucket %s initialized successfully", bucketName));
```
Comment (rmatharu) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: can be simplified into:
LOGGER.info("Read function for bucket {} initialized successfully", bucketName);

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableWriteFunction.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    super(bucketName, clusterNodes, valueClass);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info(String.format("Write function for bucket %s initialized successfully", bucketName));
```
Comment (rmatharu) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableWriteFunction.java]: LOGGER.info("Write function for bucket {} initialized successfully", bucketName);

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
+  protected Duration timeout = Duration.ZERO; // default value 0 means no timeout
+  protected Duration ttl = Duration.ZERO; // default value 0 means no ttl, data will be stored forever
```
Comment (rmatharu) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: These defaults are a bit counter-intuitive. 
You can use java's Optional<Duration> on timeout and ttl to mean that these features are enabled, else disabled.

Alternately, would it be possible for a negative value of timeout to mean "no timeout", similarly for ttl.
Duration does allow negative.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 1 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: Thank you Raymond for review.
There is one case when the valueType is JsonObject.class, we do not require a valueSerde. The (de)serialization is handled by Couchbase client in that case. That's why we don't enforce valueSerde to be not null.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 2 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: This method is inherited by the child classes and is required by TableReadFunction and TableWriteFunction. Although it's not required by any interface in BaseCouchbaseTableFunction class, we cannot change its signature.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java (hunk 3 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: Sounds good.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 5 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: These defaults are consistent with Couchbase client. Actually, even if I don't set ttl, it will still be 0 in a Couchbase document. So I use those default values to be consistent with Couchbase client API. Otherwise, we need to add extra logic to check those values in read/write function, making it a little more complicated. How do you think?

Comment (rmatharu) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: I see, in that case, if you can please add the 0-special-case behavior to the javadoc for methods withTimeout() and withTtl().

File: samza-kv-remote/src/test/java/org/apache/samza/table/remote/couchbase/TestCouchbaseBucketRegistry.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+import org.junit.runner.RunWith;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import static org.junit.Assert.*;
+import static org.mockito.Matchers.*;
+import static org.powermock.api.mockito.PowerMockito.*;
+
+
+@RunWith(PowerMockRunner.class)
```
Comment (rmatharu) [inline samza-kv-remote/src/test/java/org/apache/samza/table/remote/couchbase/TestCouchbaseBucketRegistry.java]: If you can add a brief description of the class here, similar to how you've added for TestCouchbaseRemoteTableEndToEnd. 
It will greatly help for future development, when a change in code breaks this test, and requires someone to debug the test breakage.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 5 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: No problem.

File: samza-kv-remote/src/test/java/org/apache/samza/table/remote/couchbase/TestCouchbaseBucketRegistry.java (hunk 6 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/test/java/org/apache/samza/table/remote/couchbase/TestCouchbaseBucketRegistry.java]: Good point.

**Comment by Zhangyx39** [comment]
> Minor comments, feel free to fix and checkin.

Thanks for the review @rmatharu 
But I don't have permission to checkin the code. Can you help me to do that?

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+  /**
+   * Constructor for BaseCouchbaseTableFunction. This constructor abstracts the shareable logic of the read and write
+   * functions. It is not intended to be called directly.
+   * @param bucketName Name of the Couchbase bucket
+   * @param clusterNodes Some Hosts of the Couchbase cluster. Recommended to provide more than one nodes so that if
+   *                     the first node could not be connected, other nodes can be tried.
+   * @param valueClass type of values
+   */
+  public BaseCouchbaseTableFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: Yep. This could be nicer.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+   * @param password password
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withUsernameAndPassword(String username, String password) {
+    if (environmentConfigs.sslEnabled != null && environmentConfigs.sslEnabled) {
+      throw new IllegalArgumentException(
+          "Role-Based Access Control and Certificate-Based Authentication cannot be used together.");
+    }
+    environmentConfigs.username = username;
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: I considered about it twice. We can add check for username. But we shouldn't enforce the password to be not empty because there could be some buckets migrated from pre 5.0 Couchbase versions that has no password. In that case, simple an username without password should be able to access the bucket.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+  /**
+   * Enable certificate-based authentication. If ssl is enabled sslKeystore or sslTrustStore should also be provided
+   * accordingly.
+   * @param sslEnabled allows to enable certificate-based authentication
+   * @param certAuthEnabled allows to enable X.509 client certificate authentication
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withSslEnabledAndCertAuthEnabled(boolean sslEnabled,
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: There are four combinations here:

- Both true, enable certificate-based authentication
- Both false, disable
- sslEnabled: true, certAuthEnabled: false, only server side certificate checking is needed
- sslEnabled: false, certAuthEnabled: true, not allowed, which is checked in line 176

File: build.gradle [hunk 10]
```diff
@@ -697,6 +698,34 @@ project(":samza-kv-rocksdb_$scalaSuffix") {
   }
 }
 
+project(":samza-kv-remote_$scalaSuffix") {
```
Comment (prateekm) [inline build.gradle]: Should this be samza-kv-couchbase?

File: build.gradle [hunk 11]
```diff
@@ -188,6 +188,7 @@ project(":samza-core_$scalaSuffix") {
     compile "org.scala-lang:scala-library:$scalaVersion"
     compile "org.slf4j:slf4j-api:$slf4jVersion"
     compile "net.jodah:failsafe:$failsafeVersion"
+    compile "org.scala-lang:scala-library:$scalaVersion"
```
Comment (prateekm) [inline build.gradle]: Why is this required?

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    if (configs.bootstrapHttpSslPort != null) {
+      envBuilder.bootstrapHttpSslPort(configs.bootstrapHttpSslPort);
+    }
+    CouchbaseEnvironment env = envBuilder.build();
+    Cluster cluster = CouchbaseCluster.create(env, clusterNodes);
+    if (configs.sslEnabled != null && configs.sslEnabled) {
+      cluster.authenticate(CertAuthenticator.INSTANCE);
+    } else if (configs.username != null && configs.password != null) {
+      cluster.authenticate(configs.username, configs.password);
+    }
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: I think we can log a warning here since in some situations we don't need authentication. E.g. mocked couchbase cluster.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+   * {@inheritDoc}
+   */
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info("Read function for bucket {} initialized successfully", bucketName);
+  }
+
+  /**
+   * {@inheritDoc}
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: I see. Then I think we can remove these.

File: build.gradle (hunk 10 above)
Comment (Zhangyx39) [reply build.gradle]: We are thinking to use samza-kv-remote as the package for all remote table providers.

File: build.gradle (hunk 11 above)
Comment (Zhangyx39) [reply build.gradle]: This could be a mistake when I rebased the branch with latest master.

File: build.gradle (hunk 10 above)
Comment (prateekm) [reply build.gradle]: Why? Is there something shared across all providers? If so, does it makes sense to separate them into two packages - samza-kv-remote and samza-kv-couchbase? If we add another remote table provider in the future, it doesn't make sense to require users to pull in couchbase dependencies as well.

Comment (Zhangyx39) [reply build.gradle]: What you said made sense. What I thought was that keeping the remote table providers together makes it easy for users to find them. But I don't have strong opinion on this since I am not familiar with how Samza organizes modules. @weisong44 How do you think about it?

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  /**
+   * A synchronized method to open a bucket given the bucket name and cluster nodes. Once a bucket has been opened,
+   * the reference of the bucket will be stored in the registry, and the same reference will be returned given same
+   * bucket name. Each time this method is called, the counter of usage of the bucket is increased by one.
+   * @param bucketName name of the bucket to be opened
+   * @param clusterNodes list of cluster nodes
+   * @param configs couchbase environment configs
+   * @return A Bucket instance associated with the bucket name and cluster nodes
+   */
+  public synchronized Bucket getBucket(String bucketName, List<String> clusterNodes,
```
Comment (mynameborat) [inline samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: Do you need synchronized here? Would concurrent map work instead?

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+
+
+/**
+ * The CouchbaseBucketRegistry is intended to reuse the same {@link Cluster} instance given same clusterNodes and reuse
+ * the same {@link Bucket} given same bucketName. Instantiating a Bucket is expensive. Different tasks within a
+ * container that is using the same bucket should use this registry to avoid creating multiple Buckets.
+ */
+public class CouchbaseBucketRegistry {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CouchbaseBucketRegistry.class);
+  private Map<String, Bucket> openedBuckets;
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: Sure. That's more appropriate.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableWriteFunction.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public void init(Context context) {
+    super.init(context);
+    LOGGER.info("Write function for bucket {} initialized successfully", bucketName);
+  }
+
+  @Override
+  public CompletableFuture<Void> putAsync(String key, V record) {
+    Preconditions.checkNotNull(key);
+    Preconditions.checkArgument(!key.contains(" "), "key must not contain spaces.");
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableWriteFunction.java]: Right. Empty string is also not valid. Should check this.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java (hunk 14 above)
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: We do need synchronized here. The purpose is that we not only want to keep the map correct, but we also want to prevent creating multiple Bucket instances in different thread. Think about this:
1. Thread 1 and thread 2 both check that the "bucket-id" key does not exist in the map.
2. Then they both try to create a new bucket instance and put to the map
3. Although the map is thread safe and only one Bucket instance can be put to the map with key "bucket-id", two instances of Bucket have been created.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: Good, so if someone calls this method, sslEnabled must be true. So it doesn't need to be a parameter, correct?

File: build.gradle (hunk 10 above)
Comment (weisong44) [reply build.gradle]: Yeah makes sense, let's make the change.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: Right. Let's remove it.

File: samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+  protected String sslTruststoreFile;
+  protected String sslTruststorePassword;
+  protected Integer bootstrapCarrierDirectPort;
+  protected Integer bootstrapCarrierSslPort;
+  protected Integer bootstrapHttpDirectPort;
+  protected Integer bootstrapHttpSslPort;
+  protected String username;
+  protected String password;
+
+  public Boolean getSslEnabled() {
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java]: I didn't mean to make it an inner class in CouchbaseEnvironmentConfigs, it makes it odd. Can we just keep it a separate class, just not use public in class declaration?

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+   *                     the first node could not be connected, other nodes can be tried.
+   */
+  public BaseCouchbaseTableFunction(String bucketName, Class<V> valueClass, String... clusterNodes) {
+    Preconditions.checkArgument(StringUtils.isNotEmpty(bucketName), "Bucket name is not allowed to be null or empty.");
+    Preconditions.checkArgument(valueClass != null, "Value class is not allowed to be null.");
+    Preconditions.checkArgument(ArrayUtils.isNotEmpty(clusterNodes),
+        "Cluster nodes is not allowed to be null or empty.");
+    this.bucketName = bucketName;
+    this.clusterNodes = ImmutableList.copyOf(clusterNodes);
+    this.environmentConfigs = new CouchbaseEnvironmentConfigs();
```
Comment (dengpanyin) [inline samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: only use "this" when necessary

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+    }
+    return bucketClosed && clusterClosed;
+  }
+
+  /**
+   * Helper method to open a cluster given cluster nodes and environment configurations.
+   */
+  private Cluster openCluster(List<String> clusterNodes, CouchbaseEnvironmentConfigs configs) {
+    DefaultCouchbaseEnvironment.Builder envBuilder = new DefaultCouchbaseEnvironment.Builder();
+    if (configs.sslEnabled != null) {
```
Comment (dengpanyin) [inline samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: do you need all the null check?

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+
+/**
+ * The CouchbaseBucketRegistry is intended to reuse the same {@link Cluster} instance given same clusterNodes and reuse
+ * the same {@link Bucket} given same bucketName. Instantiating a Bucket is expensive. Different tasks within a
+ * container that is using the same bucket should use this registry to avoid creating multiple Buckets.
+ */
+public class CouchbaseBucketRegistry {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CouchbaseBucketRegistry.class);
+  private final Map<String, Bucket> openedBuckets;
```
Comment (dengpanyin) [inline samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: probably a ConcurrentHashMap can satisfy the thread-safe-requirement instead of using "synchronized" keyword to the method.

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote.couchbase;
+
+import java.io.Serializable;
+
+
+class CouchbaseEnvironmentConfigs implements Serializable {
```
Comment (dengpanyin) [inline samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java]: Why creating this class other than use DefaultCouchbaseEnvironment.Builder  directly?

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+  @Override
+  public CompletableFuture<V> getAsync(String key) {
+    Preconditions.checkArgument(StringUtils.isNotBlank(key), "key must not be null, empty or blank");
+    CompletableFuture<V> future = new CompletableFuture<>();
+    Single<? extends Document<?>> singleObservable =
+        bucket.async().get(key, documentType, timeout.toMillis(), TimeUnit.MILLISECONDS).toSingle();
+    singleObservable.subscribe(new SingleSubscriber<Document<?>>() {
+      @Override
+      public void onSuccess(Document<?> document) {
+        if (document != null) {
```
Comment (dengpanyin) [inline samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: when will the document be null?

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+   * X.509 client cert authentication you might as well just use {@link #withSslTruststoreFileAndPassword}
+   * alone.
+   * @param sslKeystoreFile  path of ssl keystore file
+   * @param sslKeystorePassword password of ssl keystore
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withSslKeystoreFileAndPassword(String sslKeystoreFile,
+      String sslKeystorePassword) {
+    Preconditions.checkArgument(StringUtils.isNotEmpty(sslKeystoreFile), "Null or empty sslKeystoreFile");
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: The Couchbase client is using `sslKeystoreFile` so I want to keep the name the same. I guess they think keystore is a word?

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java (hunk 18 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: ok

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java (hunk 19 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: I think so because we don't have default values for the environment configs. We can't provide default values because it depends on use cases. So it's better to let user set it when they need to.

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java (hunk 21 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseEnvironmentConfigs.java]: Good question. I really wish to use DefaultCouchbaseEnvironment.Builder. But the problem is that class is not serializable. So we have to create this serializable CouchbaseEnvironmentConfigs to keep the configs.

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java (hunk 22 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: When the key doesn't exist in Couchbase. Instead of returning a document with null value, it will return a null.

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java (hunk 20 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: We not only want to keep it thread safe here, we also want to prevent multiple calls of `cluster.openBucket()`. See my previous reply:
https://github.com/apache/samza/pull/986#discussion_r276424654

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+  protected Serde<V> valueSerde = null;
+  protected Duration timeout = Duration.ZERO; // default value 0 means no timeout
+  protected Duration ttl = Duration.ZERO; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected final List<String> clusterNodes;
+  protected final String bucketName;
+
+  // Environment Settings
+  protected CouchbaseEnvironmentConfigs environmentConfigs;
```
[the remark this answers is not in the document]
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java]: sure

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java (hunk 22 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java]: I see. The if it returns a null, it will go to the onError method. In onSuccess method, a document should not be null.

File: samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java (hunk 20 above)
Comment (Zhangyx39) [reply samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java]: I see. We can use ConcurrentHashMap.computeIfAbsent can do the job. But it makes it harder to count the usage of a bucket. Since this method will only be called once per task per table, it's ok to use a synchronized method.
